### DISCUSSION TITLE
Is a Permissive License Available?

### DISCUSSION THREAD

**Discussion by romainmenke** [Myndex/apca-w3#12] [General]
Hi 👋 

I maintain `postcss-preset-env` and a bunch of PostCSS plugins.

examples of what I do :
- https://preset-env.cssdb.org/blog/css-color-parser-v1.0.0/
- https://preset-env.cssdb.org/playground/

I would like to create a PostCSS plugin for the `contrast-color(<color>)` CSS function (currently being specced). This CSS function returns either `black` or `white` and has 1 single argument of type `color`. The given color is treated as the background color and the returned color as the text color. I would like to adopt `apca-w3` for this feature.

But I have some concerns about the current license.

-------------

Part of open source is that open source software can be used by anyone for anything without exception. But that isn't true for `apca-w3`.

I think the intention behind the current license text was :
- prevent false advertising by 3rd parties
- prevent the perception of issues with the underlying algorithm when in reality a given tool is using an outdated version

(correct me if I am wrong here :) )

But this is highly problematic for tool creators.

However I can't adopt `apca-w3` because it's impossible for me to adhere to it's license.
- I don't know how my tools will be used by others
- I definitely can't adopt a license that would give anyone the right to audit the software of users of my tools
- I don't know when I will have time to do maintenance
- I might get hit by a bus, in which case I will also no longer be able to do maintenance

This license might be ok for W3 work, research and usage by a large organisation with substantial resources, but it can't work for open source tools.

The alternative to use an algorithm that isn't as good.

------

Would you be open to discussing some alternatives?

We really like the results of `apca-w3` and what this could mean for end users who really need web content to have good contrast.

------

Alternatives :

- a release of this package with an [osi approved license](https://opensource.org/licenses/)
- a one-off [osi approved license](https://opensource.org/licenses/) for my use-case only. This would allow me to embed a copy of `apca-w3`. (it would not allow me to distribute the algorithm as a standalone package)
- ...

**Comment by Myndex** [comment]
Hi Roman @romainmenke, if you are only doing a "font flipper" that flips text #000 or #fff, then there are alternates with less code than APCA. These are all AGPL 3:

There is [DPS contrast,](https://github.com/Myndex/deltaphistar) and I discuss flipping at the [font-flip repo](https://github.com/Myndex/fancyfontflipping).

## _But wait there's more_
If you want an **APCA compatible flipper**, I just created one for you, and here's the repo:

### [max-contrast](https://github.com/Myndex/max-contrast)<br>_an APCA-compatible B&W flipper_

```JS

    let textColor = maxContrast(bgColor); // returned color is black or white
```

There's two files, one sends and returns and RGB array, the other file sends int tuples and returns a color string. AGPL 3 license, if you want I could publish it as a package, but if you take a look you'll see it's pretty simple. Please let me know if you have any questions.

If you decide you need _more complete contrast_ using APCA, I can help you with a license for that too. But if you only wanna flip black / white, this is the way to do it.

Thank you for reading,

Andy

https://github.com/Myndex/max-contrast

**Reply by romainmenke** [reply]
Hi @Myndex,

This looks awesome and thank you so much for your effort in putting this together!

Unfortunately AGPL 3 is a copyleft license which is also incompatible with our project. Although it is OSI approved, which is what I asked for. I didn't know that not everything from the OSI approved list was compatible with our project at the time of asking.

Everything we do is MIT-0 licensed and using algorithms that are AGPL 3 licensed would effectively make our project also AGPL 3 because it is all executed as a single program. It might also force our users to open source parts or all of their work, which is not something our license requires today.

This is obviously your choice and I fully respect this :)
I also see the benefits of AGPL 3 for your work at this time.

Our project provides polyfills and fallbacks for older browser versions.
Given that eventually all users will update to a modern browser version, our code and project naturally loses it's value over time.

Because of this aspect we have a very permissive license.
We embrace the fact that, in the end, our code has zero value.

The same isn't true for what you are doing with APCA.
So I don't want to put you into a position to change to a license you do not want to adopt for our benefit.

I obviously hope that a more permissive (non-copyleft) license could be possible, now or in the future. But I respect and understand any choice you make in this matter.

------

Thank you again for all your hard work in this area!

**Reply by Myndex** [reply]
Can you work with GNU General Public License? Or what licenses can you work with?

**Reply by romainmenke** [reply]
The most common licenses in our dependencies today are :
- 0BSD
- Apache-2.0
- BSD-2-Clause
- BSD-3-Clause
- CC-BY-3.0
- CC-BY-4.0
- CC0-1.0
- ISC
- MIT
- MIT-0

I've mostly assigned the MIT license to my own projects.

Any [permissive license](https://snyk.io/learn/open-source-licenses/#permissive) would be fine for our use case.

----------

But as I said before, I fully understand if such a license isn't compatible with your work at this time.

**Comment by Myndex** [comment]
Hi @romainmenke  just to answer your other questions:

> _I think the intention behind the current license text was:_
> - _prevent false advertising by 3rd parties_
> - _prevent the perception of issues with the underlying algorithm when in reality a given tool is using an outdated version_
> _(correct me if I am wrong here :) )_

Yes, you are correct, because there already have/had been some bad actors stirring up problems, including versions that were not at all APCA, and creating issues in bad faith, apparently to specifically harass. It's ridiculous like a Monty Python sketch.

APCA represents an important paradigm shift, and such changes to the gestalt are always met by usurpers, miscreants, and your regular, run-of-the-mill, talentless hacks. 

> _But this is highly problematic for tool creators._

I know, and I am sorry about that. This is a temporary situation during the public beta, and will not be permanent, and **_there will be a free-to-use library with a permissive license._**

In the meantime tell me if the font flip I gave you above is sufficient for your current needs, and if you need more, we'll get you sorted.


Thank you for reading.

**Reply by Ariane-B** [reply]
> I know, and I am sorry about that. This is a temporary situation during the public beta, and will not be permanent, and there will be a free-to-use library with a permissive license.

Hi again, a few months later! Do you have an ETA of when a library usable by closed-source apps will come out?

For now we're using [colord](https://github.com/omgovich/colord) and it seems to work fine, but apca-w3 seemed superior to me, so it feels like a bit of a shame.

**Comment by Myndex** [comment]
[![RadMax being chased down by a belligerent dune buggy in the desert, and big letters in the sand are MAX, and under that says CONTRAST](https://github.com/Myndex/apca-w3/assets/42009457/f1818396-ffee-4e15-b359-1af001bc957d)](https://github.com/Myndex/max-contrast)
